lib/fswatcher/fswatcher.go

```diff
@@ -0,0 +1,558 @@
+	// If a dir existed at path, it would be removed from dirs, thus
+	// childCount would not increase.
+	if !ok && parentDir.childCount() == localMaxFilesPerDir {
+		l.Debugf("%v Parent dir already has %d children, tracking it instead: %s", w, localMaxFilesPerDir, evPath)
+		w.aggregateEvent(filepath.Dir(evPath), evTime, evType)
+		return
+	}
+
+	firstModTime := evTime
+	if ok {
+		firstModTime = childDir.getFirstModTime()
+		evType |= childDir.getEventType()
+		delete(parentDir.dirs, evPath)
+	}
+	l.Debugf("%v Tracking (type %v): %s", w, evType, evPath)
+	parentDir.events[evPath] = &event{
+		path:         evPath,
+		firstModTime: firstModTime,
+		lastModTime:  evTime,
+		evType:       evType,
+	}
+	w.resetNotifyTimerIfNeeded()
+}
+
+func (w *watcher) actOnTimer() {
+	eventCount := w.rootEventDir.eventCount()
+	if eventCount == 0 {
+		l.Debugln(w, "No tracked events, waiting for new event.")
+		w.notifyTimerNeedsReset = true
+		return
+	}
+	oldFsEvents := w.popOldEvents(w.rootEventDir, time.Now())
+	if len(oldFsEvents) == 0 {
+		l.Debugln(w, "No old fs events")
+		w.resetNotifyTimer(w.notifyDelay)
+		return
+	}
+	// Sending to channel might block for a long time, but we need to keep
+	// reading from notify backend channel to avoid overflow
+	go func() {
+		timeBeforeSending := time.Now()
+		l.Debugf("%v Notifying about %d fs events", w, len(oldFsEvents))
+		separatedBatches := make(map[eventType][]string)
+		for path, event := range oldFsEvents {
+			separatedBatches[event.evType] = append(separatedBatches[event.evType], path)
+		}
+		for _, evType := range [3]eventType{nonRemove, mixed, remove} {
+			if len(separatedBatches[evType]) != 0 {
+				select {
+				case w.notifyChan <- separatedBatches[evType]:
+				case <-w.stop:
+					return
+				}
+			}
+		}
+		// If sending to channel blocked for a long time,
+		// shorten next notifyDelay accordingly.
+		duration := time.Since(timeBeforeSending)
+		buffer := time.Duration(1) * time.Millisecond
+		var nextDelay time.Duration
+		switch {
+		case duration < w.notifyDelay/10:
+			nextDelay = w.notifyDelay
+		case duration+buffer > w.notifyDelay:
+			nextDelay = buffer
+		default:
+			nextDelay = w.notifyDelay - duration
+		}
+		select {
+		case w.notifyTimerResetChan <- nextDelay:
+		case <-w.stop:
+		}
+	}()
+	return
+}
+
+// popOldEvents removes events that should be sent to
+func (w *watcher) popOldEvents(dir *eventDir, currTime time.Time) eventBatch {
+	oldEvents := make(eventBatch)
+	for _, childDir := range dir.dirs {
+		for path, event := range w.popOldEvents(childDir, currTime) {
+			oldEvents[path] = event
+		}
+	}
+	for path, event := range dir.events {
+		// 2 * results in mean event age of notifyDelay (assuming randomly
+		// occurring events). Reoccuring and remove/mixed events
+		// (for efficient renames) events are delayed until notifyTimeout.
+		if (event.evType == nonRemove && 2*currTime.Sub(event.lastModTime) > w.notifyDelay) || currTime.Sub(event.firstModTime) > w.notifyTimeout {
```

Comment:
I will move this entire ugly condition into it's own function `isOld` and explained more extensively about what and why. Please review that and bring it up again if it's still not clear.